### PR TITLE
Enable asset transformation feature in hot reload workflow (excluding Web)

### PR DESCRIPTION
**Pull request by andrewkolos**
Partial implementation of https://github.com/flutter/flutter/issues/143348

This enables asset transformation during hot reload (except for web, because that has its own implementation of `DevFS` 🙃). Asset transformers will be reapplied after changing any asset and performing a hot reload during `flutter run`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:187]: We end up doing a redundant copy of the initial input file here, but it's unclear if it's worth having a boolean flag parameter or a separate method on `AssetTransformer` that transforms the asset without doing the defensive copy. Alternatively, the code calling `AssetTransformer::transformAsset` can do the defensive copy themselves, but that would be even more copy-paste code in all the places that write assets.

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:205]: Oh, I suppose since we're not catching exceptions, a thrown exception would bubble up and avoid us reaching this line?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:205]: ok, nvm, this seems safe

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:174]: `late` should generally be avoided (except it's useful in test code, where you ensure it's always initialized from `setUp()`). Consider if line 172 throws and we never initialize `inputFile`, we'd get a `LateInitializationError` on line 202. Maybe we make it nullable, and then update the check on 201 to be `if (inputFile != null && cleanupInput)`?

I think it's the right tool for the job here of unconditionally cleaning up resources, however `finally` can make it more difficult to reason about the type safety of a block of code.

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:187]: Yeah, I'm leaning towards keeping the code simpler. this LGTM.

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:180]: I'm not familiar with the implementations of `DevFSContent`. When would we get `DevFSFileContent` vs say `DevFSByteContent`?

File: packages/flutter_tools/lib/src/devfs.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/devfs.dart:711]: for my understanding, why do we not evict this path the first upload?

**Review by christopherfujino — DISMISSED** [review]
2 nits, about avoiding `late` and `await`-ing a future, but otherwise LGTM

**Comment by andrewkolos** [comment]
One issue worth pointing out is that if (during `flutter run`) 
1) an asset's transformers are changed and 
2) there is a hot reload

then the asset won't be re-transformed. This is because the only thing we [check](https://github.com/andrewkolos/flutter/blob/615b9ea6c1aeaaeeca4ab997eb5ad4772857eaa4/packages/flutter_tools/lib/src/devfs.dart#L642) during a DevFS update is the last-modified time of the asset file.

**Comment by christopherfujino** [comment]
> One issue worth pointing out is that if (during `flutter run`)
> 
> 1. an asset's transformers are changed and
> 2. there is a hot reload
> 
> then the asset won't be re-transformed. This is because the only thing we [check](https://github.com/andrewkolos/flutter/blob/615b9ea6c1aeaaeeca4ab997eb5ad4772857eaa4/packages/flutter_tools/lib/src/devfs.dart#L642) during a DevFS update is the last-modified time of the asset file.

What is the "asset file" in this context? the output of the transformer?

And is this a temporary bug, or a known limitation that we should document?

**Comment by andrewkolos** [comment]
> > One issue worth pointing out is that if (during `flutter run`)
> > 
> > 1. an asset's transformers are changed and
> > 2. there is a hot reload
> > 
> > then the asset won't be re-transformed. This is because the only thing we [check](https://github.com/andrewkolos/flutter/blob/615b9ea6c1aeaaeeca4ab997eb5ad4772857eaa4/packages/flutter_tools/lib/src/devfs.dart#L642) during a DevFS update is the last-modified time of the asset file.
> 
> What is the "asset file" in this context? the output of the transformer?
> 
> And is this a temporary bug, or a known limitation that we should document?

The original file. This is a limitation of how we implement hot reload. More research and another PR likely to follow.

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:181]: I wouldn't be surprised if this is always a `DevFSFileContent`. However, this code can't know that. There is probably a missing abstraction somewhere upstream that guarantees this is a file, but I don't improving this situation would be easy.

File: packages/flutter_tools/lib/src/devfs.dart
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/devfs.dart:696]: The latter is the only option because adding `await` will trigger the unnecessary `await` lint.

File: packages/flutter_tools/lib/src/devfs.dart
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/devfs.dart:712]: Good question. The only reference to `assetsPathToEvict` is here: https://github.com/flutter/flutter/blob/4e70bfae2bff2c272af7fb7a870695edd9f8cfb4/packages/flutter_tools/lib/src/run_hot.dart#L1207-L1215

where make some vm service call, `ext.flutter.evict`. Do you know where to find the code that process this call?

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/devfs.dart:712]: Nevermind. I'm starting to figure this out, give me a sec.

File: packages/flutter_tools/lib/src/devfs.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/devfs.dart:696]: o rly? that sounds like a bug in the lint, given the bug I linked.

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/devfs.dart:696]: For any future historians: here's some context https://github.com/dart-lang/language/issues/870#issuecomment-1265066268. We'll try to follow up on getting this lint adjusted in the meantime.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/devfs.dart:696]: I'll follow up on the lint, you can leave this as is or make it sync

File: packages/flutter_tools/lib/src/devfs.dart
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/devfs.dart:712]: So this VM service call ends up here: https://github.com/flutter/flutter/blob/7a4c2465afe875355aaac3efee158084f13e6fc7/packages/flutter/lib/src/services/binding.dart#L250-L258

which indeed just evicts the asset from the framework's cache. However, I don't know if there is a specific reason to _not_ vacuously evict from the cache in first upload (other than perhaps performance).

**Comment by andrewkolos** [comment]
I believe I've fixed the hot reload issue. Changes are in this commit: [fix asset not being updated when transformers are changed before hot reload](https://github.com/flutter/flutter/pull/144161/commits/49cb3f03ee083e86f492d762b846c74e656a0c31).

This works primarily by utilizing the previously dormant [`bundleDirty` parameter on `ResidentRunner::updateDevFS`](https://github.com/flutter/flutter/blob/42b02d0a3a0def43cfc246593a7631249c25c03b/packages/flutter_tools/lib/src/resident_runner.dart#L561). However, this means that, [if the asset bundle becomes dirty for whatever reason](https://github.com/flutter/flutter/blob/49cb3f03ee083e86f492d762b846c74e656a0c31/packages/flutter_tools/lib/src/asset.dart#L215), we'll end up re-writing _all assets_.

I am not sure this is acceptable, as any modification to a pubspec basically means all assets needs to be processed and rewritten to the device, which could be slow given the right circumstances (or basically any case with transformers). I want to experiment with more granular caching, though this will likely make all this code even more complicated than it already is.

**Comment by andrewkolos** [comment]
> I believe I've fixed the hot reload issue. Changes are in this commit: [fix asset not being updated when transformers are changed before hot reload](https://github.com/flutter/flutter/pull/144161/commits/49cb3f03ee083e86f492d762b846c74e656a0c31).
> 
> This works primarily by utilizing the previously dormant [`bundleDirty` parameter on `ResidentRunner::updateDevFS`](https://github.com/flutter/flutter/blob/42b02d0a3a0def43cfc246593a7631249c25c03b/packages/flutter_tools/lib/src/resident_runner.dart#L561). However, this means that, [if the asset bundle becomes dirty for whatever reason](https://github.com/flutter/flutter/blob/49cb3f03ee083e86f492d762b846c74e656a0c31/packages/flutter_tools/lib/src/asset.dart#L215), we'll end up re-writing _all assets_.
> 
> I am not sure this is acceptable, as any modification to a pubspec basically means all assets needs to be processed and rewritten to the device, which could be slow given the right circumstances (or basically any case with transformers). I want to experiment with more granular caching, though this will likely make all this code even more complicated than it already is.

I thought about this more and brought this up during this week's tool PR review. I'm OK with punting on this by documenting this feature as not being supported in hot-reload. To make this feature work with hot reload, we'd need to do some significant refactoring to asset bundling+writing code (to avoid adding more spaghetti than there already is). We need to overhaul what makes an asset "dirty" so we know when to retransform and rewrite it to the device. Currently, this mechanism is implemented as a function of the last-modified-time on the underlying asset file (or for non-files, [a boolean that returns true the first time it's read and then false afterwards](https://github.com/flutter/flutter/blob/37c8df1690ffb0e41877aa4e52191707aa056a5a/packages/flutter_tools/lib/src/devfs.dart#L166-L172)). It is unclear if this is worth overhauling at this time, but perhaps such a decision would be easier if we had another asset related feature that would require a similar refactoring to support hot reload.

**Comment by andrewkolos** [comment]
@christopherfujino, after this PR lands, I'll make a documentative PR that demonstrates a duct-tape/hacky solution to make this work for hot reload. It'll include comments that point out the obstacles in the way of making hot reload work well for this. This could be referenced in the future when we come back to this.

**Comment by christopherfujino** [comment]
> @christopherfujino, after this PR lands, I'll make a documentative PR that demonstrates a duct-tape/hacky solution to make this work for hot reload. It'll include comments that point out the obstacles in the way of making hot reload work well for this. This could be referenced in the future when we come back to this.

SGTM

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart]: Addressed in [nit](https://github.com/flutter/flutter/pull/144161/commits/37bb741ba64a22b957d1ef1e7ded27c7e7ea11ad), but in slightly a different way than proposed. I make a null-assertion on `inputFile`, since it should be set if `cleanupInput` was set.

**Comment by andrewkolos** [comment]
@christopherfujino poke

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:202]: it could be null line 179 `resource = await _transformationPool.request();` throws, right?

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart]: I've added a null check for extra safety.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by andrewkolos** [comment]
More thoughts about hot-reload:

Another challenge with making this fully-support hot reload is that `dart run <package>` seems prone to failing when running concurrently with a `pub get`:

```
User-defined transformation of asset "/Users/andrewkolos/Desktop/asset_transformers_test/assets/svg.svg" failed.
Transformer process terminated with non-zero exit code: 255
Transformer package: vector_graphics_transformer
Full command: /Users/andrewkolos/Documents/GitHub/flutter/bin/cache/dart-sdk/bin/dart run vector_graphics_transformer --input=/var/folders/qh/m_ckdxls0xvfc7ccjj13yjjw010s5h/T/flutter_tools.f7R8Z7/svg.svg-transformOutput0.svg
--output=/var/folders/qh/m_ckdxls0xvfc7ccjj13yjjw010s5h/T/flutter_tools.f7R8Z7/svg.svg-transformOutput1.svg
stdout:

stderr:
Unhandled exception:
PathNotFoundException: Cannot open file, path = '/Users/andrewkolos/Desktop/asset_transformers_test/.dart_tool/pub/bin/vector_graphics_transformer/tmpKco51R/vector_graphics_transformer.dart-3.4.0-187.0.dev.snapshot.incremental.temp' (OS Error: No such file or directory, errno = 2)
#0      _checkForErrorResponse (dart:io/common.dart:55)
#1      _File.open.<anonymous closure> (dart:io/file_impl.dart:381)
<asynchronous suspension>
#2      _FileStreamConsumer.addStream.<anonymous closure> (dart:io/file_impl.dart:205)
<asynchronous suspension>
Compilation did not produce any result. Expected file at `/Users/andrewkolos/Desktop/asset_transformers_test/.dart_tool/pub/bin/vector_graphics_transformer/tmpKco51R/vector_graphics_transformer.dart-3.4.0-187.0.dev.snapshot.incremental.temp`
Cannot rename file to '/Users/andrewkolos/Desktop/asset_transformers_test/.dart_tool/pub/bin/vector_graphics_transformer/vector_graphics_transformer.dart-3.4.0-187.0.dev.snapshot', path =
'/Users/andrewkolos/Desktop/asset_transformers_test/.dart_tool/pub/bin/vector_graphics_transformer/tmpKco51R/vector_graphics_transformer.dart-3.4.0-187.0.dev.snapshot.incremental.temp' (OS Error: No such file or directory, errno = 2)
PathNotFoundException: Cannot rename file to '/Users/andrewkolos/Desktop/asset_transformers_test/.dart_tool/pub/bin/vector_graphics_transformer/vector_graphics_transformer.dart-3.4.0-187.0.dev.snapshot', path =
'/Users/andrewkolos/Desktop/asset_transformers_test/.dart_tool/pub/bin/vector_graphics_transformer/tmpKco51R/vector_graphics_transformer.dart-3.4.0-187.0.dev.snapshot.incremental.temp' (OS Error: No such file or directory, errno = 2)
```

This is problematic because the Dart VSCode extension seems to run `flutter pub get` automatically whenever pubspec.yaml is modified. I obtained the above error message by 1) I adding a new asset with a transformer to my pubspec, 2) saving the file, and then 3) quickly (within a few seconds) initiating a hot reload.

**Pull request merged**
